File: 03-poo/15-diseno-orientado-a-objetos/scala/soluciones/RetoResponsabilidades.scala
class ValidadorReto:
  def correoValido(c: String): Boolean = c != null && c.contains("@")
trait NotificadorReto:
  def enviar(destino: String): Unit
class ConsolaReto extends NotificadorReto:
  def enviar(destino: String): Unit = println(s"Enviado a $destino")
class ServicioRegistro(validador: ValidadorReto, notificador: NotificadorReto):
  def registrar(correo: String): Unit =
    require(validador.correoValido(correo))
    notificador.enviar(correo)

@main def retoResponsabilidades(): Unit = new ServicioRegistro(new ValidadorReto,new ConsolaReto).registrar("usuario@example.com")

File: 03-poo/15-diseno-orientado-a-objetos/java/soluciones/RetoResponsabilidades.java
public class RetoResponsabilidades {
    interface Notificador { void enviar(String destino); }
    static class Validador { boolean correoValido(String c){ return c!=null && c.contains("@"); } }
    static class ConsolaNotificador implements Notificador { public void enviar(String d){ System.out.println("Enviado a "+d); } }
    static class ServicioRegistro { private final Validador validador; private final Notificador notificador; ServicioRegistro(Validador v,Notificador n){validador=v;notificador=n;} void registrar(String correo){ if(!validador.correoValido(correo)) throw new IllegalArgumentException(); notificador.enviar(correo); } }
    public static void main(String[] args){ new ServicioRegistro(new Validador(),new ConsolaNotificador()).registrar("usuario@example.com"); }
}
